### PR TITLE
Add neighbor_parallel functor tests

### LINKED ISSUE #414: Missing work_tag support for neighbor_parallel_for and neighbor_parallel_reduce
`Cabana::neighbor_parallel_for` and `Cabana::neighbor_parallel_reduce` do not support `function name tags`. 

The internal lambda has no `work_tag`
https://github.com/ECP-copa/Cabana/blob/575a692f1212a6be4059023ea75a81c25852f5c0/core/src/Cabana_Parallel.hpp#L243

However, this lambda is called via `Kokkos::parallel_for` which passes the `work_tag` as the first argument.
https://github.com/ECP-copa/Cabana/blob/575a692f1212a6be4059023ea75a81c25852f5c0/core/src/Cabana_Parallel.hpp#L253

### PR DESCRIPTION
**Pull request by streeve**
All previous neighbor parallel tests used lambdas. Our implementation works fine for team parallelism, but the explicit indexing pointed out by @XzzX breaks serial with a work tag

Fixes #414 
First part of #422

### CODE REVIEW

**Review by sslattery — APPROVED** [review]

**Comment by XzzX** [comment]
Do you want to add a test for `neighbor_parallel_reduce` with work tags as well?

**Comment by streeve** [comment]
> Do you want to add a test for `neighbor_parallel_reduce` with work tags as well?

Good point, I'll add

File: core/src/Cabana_Parallel.hpp [hunk 1]
```diff
@@ -292,6 +297,11 @@ inline void neighbor_parallel_for(
 
     using memory_space = typename neighbor_list_traits::memory_space;
 
+    auto begin = exec_policy.begin();
+    auto end = exec_policy.end();
+    using linear_policy_type = Kokkos::RangePolicy<execution_space, void, void>;
```
Comment (dalg24) [inline core/src/Cabana_Parallel.hpp:302]: Sorry I know it is a bit late but what are these two trailing `void` template parameter there for?  Does the compile with Kokkos 3.4?

Comment (streeve) [reply core/src/Cabana_Parallel.hpp:302]: Yeah those are unnecessary. It does pass with Kokkos 3.2, master, and develop though

**Pull request merged**